gloo/common/linux.cc

```diff
@@ -192,29 +192,28 @@ const std::string& infinibandToBusID(const std::string& name) {
 static int getInterfaceSpeedGLinkSettings(int sock, struct ifreq* ifr) {
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4,6,0)
   constexpr auto link_mode_data_nwords = 3 * 127;
-  struct {
-    struct ethtool_link_settings req;
-    __u32 link_mode_data[link_mode_data_nwords];
-  } ecmd;
+  constexpr auto bufsize = sizeof(struct ethtool_link_settings) + sizeof(__u32)*link_mode_data_nwords;
+  char buf[bufsize];
+  struct ethtool_link_settings& ecmd = *(struct ethtool_link_settings*)buf;
```

Comment:
IMO this is technically a strict aliasing violation and hence UB. At least `buf` needs to be aligned as `struct ethtool_link_settings` and possibly a placement new needs to be used.

Idea: Using an union to enforce the size might work:
```
  constexpr auto bufsize = sizeof(struct ethtool_link_settings) + sizeof(__u32)*link_mode_data_nwords;
  union {
    struct ethtool_link_settings req;
    char buf[bufsize];
  } ecmd;
```